monix-reactive/shared/src/test/scala/monix/reactive/internal/builders/ResourceObservableSuite.scala

```diff
@@ -0,0 +1,295 @@
+    var released = false
+    val dummy = DummyException("dummy")
+    val bracketed = Observable.resourceTask(Task.unit)(_ => Task {
+      released = true
+    })
+      .flatMap { _ =>
+        Observable.suspend[Int](Observable.raiseError(dummy))
+      }
+
+    val f = bracketed.completedL.runAsync
+    s.tick()
+
+    assertEquals(f.value, Some(Failure(dummy)))
+    assert(released)
+  }
+
+  test("Observable.resourceTask nesting: inner releases even if outer release fails") { implicit s =>
+    var released = false
+    val dummy = DummyException("dummy")
+    val bracketed = Observable.resourceTask(Task.unit)(_ => Task.raiseError(dummy))
+      .flatMap { _ =>
+        Observable.resourceTask(Task.unit)(_ => Task {
+          released = true
+        })
+          .flatMap(_ => Observable(1, 2, 3))
+      }
+
+    val f = bracketed.completedL.runAsync
+    s.tick()
+
+    assertEquals(f.value, Some(Failure(dummy)))
+    assert(released)
+  }
+
+  test("Observable.resourceTask releases resource on all completion methods") { implicit s =>
+    val rs = new Resource
+    val completes: Array[Observable[Int] => Task[Unit]] =
+      Array(
+        _.completedL,
+        _.consumeWith(Consumer.complete),
+        _.lastOrElseL(()).map(_ => ()),
+        _.findL(_ => true).map(_ => ()),
+        _.foldL.map(_ => ()),
+        _.foldWhileLeftL(())((_, _) => Left(())),
+        _.firstOrElseL(()).map(_ => ()),
+        _.forAllL(_ => true).map(_ => ()),
+        _.existsL(_ => true).map(_ => ()),
+        _.foldLeftL(())((_, _) => ()),
+        _.headOrElseL(()).map(_ => ()),
+        _.maxL.map(_ => ()),
+        _.maxByL(identity).map(_ => ()),
+        _.minL.map(_ => ()),
+        _.minByL(identity).map(_ => ()),
+        _.sumL.map(_ => ())
+      )
+
+    val pure = Observable.resourceTask(rs.acquire)(_.release)
+      .flatMap(_ => Observable(1, 2, 3))
+
+    for (method <- completes) {
+      method(pure).runAsync
+      s.tick()
+    }
+
+    assertEquals(rs.acquired, completes.length)
+    assertEquals(rs.released, completes.length)
+
+    val dummy = DummyException("dummy")
+    val faulty = Observable.resourceTask(rs.acquire)(_.release)
+      .flatMap(_ => Observable.raiseError[Int](dummy))
+
+    for (method <- completes) {
+      val f = method(faulty).runAsync
+      s.tick()
+      assertEquals(f.value, Some(Failure(dummy)))
+    }
+
+    assertEquals(rs.acquired, completes.length * 2)
+    assertEquals(rs.released, completes.length * 2)
+
+    val broken = Observable.resourceTask(rs.acquire)(_.release)
+      .flatMap(_ => Observable.suspend[Int](Observable.raiseError(dummy)))
+
+    for (method <- completes) {
+      val f = method(broken).runAsync
+      s.tick()
+      assertEquals(f.value, Some(Failure(dummy)))
+    }
+
+    assertEquals(rs.acquired, completes.length * 3)
+    assertEquals(rs.released, completes.length * 3)
+  }
+
+  test("Observable.resourceTask does not require non-strict use") { implicit s =>
```

Comment:
I took this from tests for `Iterant` but in case of `Observable` it's non-deterministic and I believe it's not something we expect to hold? Builders like `Observable.pure` call `onComplete` straight away after subscription leading to releasing resources. Or I might be understanding it wrong and it's something wrong with my implementation